src/algebraic_geometry/AffineScheme.lean

```diff
@@ -307,6 +307,125 @@ begin
+  convert PresheafedSpace.is_open_immersion.c_iso ((opens.map f.base).obj U),
+end
+
+lemma is_affine_open.from_Spec_preimage_basic_open {X : Scheme} {U : opens X.carrier}
+  (hU : is_affine_open U) (f : X.presheaf.obj (op U)) :
+  (opens.map hU.from_Spec.val.base).obj (X.basic_open f) =
+    RingedSpace.basic_open _ (Spec_Γ_identity.inv.app (X.presheaf.obj $ op U) f) :=
+begin
+  erw LocallyRingedSpace.preimage_basic_open,
+  refine eq.trans _ (RingedSpace.basic_open_res_eq (Scheme.Spec.obj $ op $ X.presheaf.obj (op U))
+    .to_LocallyRingedSpace.to_RingedSpace (eq_to_hom hU.from_Spec_base_preimage).op _),
+  congr,
+  rw ← comp_apply,
+  congr,
+  erw ← hU.Spec_Γ_identity_hom_app_from_Spec,
+  rw iso.inv_hom_id_app_assoc,
+end
+
+/-- The canonical map `Γ(𝒪ₓ, D(f)) ⟶ Γ(Spec 𝒪ₓ(U), D(Spec_Γ_identity.inv f))`
+This is an isomorphism, as witnessed by an `is_iso` instance. -/
+def basic_open_sections_to_affine {X : Scheme} {U : opens X.carrier} (hU : is_affine_open U)
+  (f : X.presheaf.obj (op U)) : X.presheaf.obj (op $ X.basic_open f) ⟶
+    (Scheme.Spec.obj $ op $ X.presheaf.obj (op U)).presheaf.obj
+      (op $ Scheme.basic_open _ $ Spec_Γ_identity.inv.app (X.presheaf.obj (op U)) f) :=
+hU.from_Spec.1.c.app (op $ X.basic_open f) ≫ (Scheme.Spec.obj $ op $ X.presheaf.obj (op U))
+  .presheaf.map (eq_to_hom $ (hU.from_Spec_preimage_basic_open f).symm).op
+
+instance {X : Scheme} {U : opens X.carrier} (hU : is_affine_open U)
+  (f : X.presheaf.obj (op U)) : is_iso (basic_open_sections_to_affine hU f) :=
+begin
+  delta basic_open_sections_to_affine,
+  apply_with is_iso.comp_is_iso { instances := ff },
+  { apply PresheafedSpace.is_open_immersion.is_iso_of_subset,
+    rw hU.from_Spec_range,
+    exact RingedSpace.basic_open_subset _ _ },
+  apply_instance
+end
+.
+lemma is_localization_basic_open {X : Scheme} {U : opens X.carrier} (hU : is_affine_open U)
+  (f : X.presheaf.obj (op U)) :
+  is_localization.away f (X.presheaf.obj (op $ X.basic_open f)) :=
+begin
+  apply (is_localization.is_localization_iff_of_ring_equiv (submonoid.powers f)
+    (as_iso $ basic_open_sections_to_affine hU f ≫ (Scheme.Spec.obj _).presheaf.map
+      (eq_to_hom (basic_open_eq_of_affine _).symm).op).CommRing_iso_to_ring_equiv).mpr,
+  convert structure_sheaf.is_localization.to_basic_open _ f,
+  change _ ≫ (basic_open_sections_to_affine hU f ≫ _) = _,
+  delta basic_open_sections_to_affine,
+  erw ring_hom.algebra_map_to_algebra,
+  simp only [Scheme.comp_val_c_app, category.assoc],
+  erw hU.from_Spec.val.c.naturality_assoc,
+  rw hU.from_Spec_app_eq,
+  dsimp,
+  simp only [category.assoc, ← functor.map_comp, ← op_comp],
+  apply structure_sheaf.to_open_res,
+end
+
+lemma basic_open_basic_open_is_basic_open {X : Scheme} {U : opens X.carrier}
+  (hU : is_affine_open U) (f : X.presheaf.obj (op U)) (g : X.presheaf.obj (op $ X.basic_open f)) :
+    ∃ f' : X.presheaf.obj (op U), X.basic_open f' = X.basic_open g :=
```

Comment:
Would it help to have a predicate `is_basic_open U f`, or something like that. Because currently we seem to need a dozen variations on a theme because sections live on opens with tricky types, etc...
Typically, it helps to have a construction (like `basic_open`) and then a characteristic predicate that describes it.

Reply:
I'm not sure what `is_basic_open U f` could do for me though. I don't have a good equivalent condition for `is_basic_open U f` except `U = X.basic_open f`, so this seems like it is just an abbreviation for that equality?